ASoC: Intel: sof_da7219_max98373: Add support for max98360a spe…

yongzhi1 opened the pull request:
…aker amp

Add Maxim MAX98360A plug-and-play Class-D amplifier support, the new card ID
is sofda7219max98360a, name is sof-da7219max98360a.

Signed-off-by: Yong Zhi <yong.zhi@intel.com>

plbossart commented on sound/soc/intel/boards/sof_da7219_max98373.c:
```diff
@@ -303,8 +331,8 @@ static struct snd_soc_dai_link dais[] = {
-static struct snd_soc_card card_da7219_m98373 = {
-	.name = "da7219max",
+static struct snd_soc_card card_da7219_max = {
+	.name = "da7219max98360a",
```
please don't do this, this is going to break existing solutions for UCM, etc.

Add a new card if needed.

yongzhi1 replied:
Ack, indeed, will fix in next drop.

plbossart commented on sound/soc/intel/boards/sof_da7219_max98373.c:
```diff
@@ -331,6 +357,18 @@ static int audio_probe(struct platform_device *pdev)
 	if (!ctx)
 		return -ENOMEM;
 
+	if (!strcmp(pdev->name, "sof_da7219_max98373")) {
```
so how is the other case handled? Or it's the default and the 98373 only adds new things on top?

yongzhi1 replied:
max98360a is set to default as it has minimal config, add more on top of it for max98373.

plbossart replied:
definitively worth a comment for other reviewers :-)

plbossart replied:
please add a comment here as discussed

plbossart commented on sound/soc/intel/common/soc-acpi-intel-jsl-match.c:
```diff
@@ -7,16 +7,22 @@
 		.sof_tplg_filename = "sof-jsl-da7219.tplg",
 	},
+	{
+		.id = "DLGS7219",
```
it'd be a better idea to use the headset codec as a constant, and the amplifier name as a quirk in all cases. You are going to reach impossible combinations otherwise.

yongzhi1 replied:
The max98360a does not have programmable interface like i2c/gpio etc, so this spk amp exists anomalously without a name, that's why I put max98373 at top.

plbossart replied:
ok, but why not follow the conventions used in the past. Just add DA7219+max98370 first, then DA7219 only, in this order. That will make it easier to deal e.g. with cases where we have a different amp.

yongzhi1 replied:
Sure, that should work.

plbossart replied:
add a comment that the order matters please, so that we don't do silly edits later on

yongzhi1 replied:
Ack, will update.

yongzhi1 commented:
Updated https://github.com/yongzhi1/linux/commit/bef2471b42bde4c489aa019e24ecdb19ce8d7d8a

rev2 update:

- Remain card_da7219_m98373 for compatibility, add card_da7219_m98360a seperately.
- Follow conventional order to add new entry in snd_soc_acpi_intel_jsl_machines.

plbossart commented on sound/soc/intel/boards/sof_da7219_max98373.c:
```diff
@@ -247,6 +271,17 @@ SND_SOC_DAILINK_DEF(idisp3_codec,
+		.dai_name = MAX98373_CODEC_DAI,
+	},
+};
+
```
don't we have a way to use the DAILINK_ macros? It's a bit ugly to hard-code when everything else uses the macros, no?

yongzhi1 replied:
I did not find a better way last time, let me double check, thanks!!

plbossart replied:
````c
SND_SOC_DAILINK_DEF(ssp1_codec,
	DAILINK_COMP_ARRAY(
	/* Left */	COMP_CODEC(MAXIM_DEV0_NAME, MAX98373_CODEC_DAI),
	/* Right */	COMP_CODEC(MAXIM_DEV1_NAME, MAX98373_CODEC_DAI)));
````
?

yongzhi1 commented:
Updated https://github.com/thesofproject/linux/commit/7e2e3e38b2d3991560a25b23be5deade925d66ea

For version 3:
- Updated comments.
- Default dai[0] to original max98373 as it's more intuitive. 
- Use SND_SOC_DAILINK_REG macro to update the dai[0] for new amp.

ranj063 commented on sound/soc/intel/boards/sof_da7219_max98373.c:
```diff
@@ -331,6 +375,16 @@ static int audio_probe(struct platform_device *pdev)
+			.no_pcm = 1,
+			.ignore_pmdown_time = 1,
+			.dpcm_playback = 1,
+			SND_SOC_DAILINK_REG(ssp0_pin, dummy) };
```
@yongzhi1 should be ssp0_pin or ssp1_pin?

yongzhi1 replied:
You are right, should be ssp1_pin, my test driver hard-coded some values, so I did not notice when doing cut & paste.

plbossart commented on sound/soc/intel/boards/sof_da7219_max98373.c:
```diff
@@ -331,6 +375,16 @@ static int audio_probe(struct platform_device *pdev)
+			.no_pcm = 1,
+			.dpcm_playback = 1,
+			.ignore_pmdown_time = 1,
+			SND_SOC_DAILINK_REG(ssp1_pin, dummy) };
```
should it be SND_SOC_DAILINK_REG(ssp1_pin, dummy, platform) ?

what you want is a dummy codec, yes?

yongzhi1 replied:
Yes, that's correct, I dropped the "platform" for aesthetic as the platform_name will be fixed up.

plbossart replied:
but not having a platform here results in no allocation of the memory, we've seen this before. the fixup will change the name, but it requires a platform to be defined initially.

yongzhi1 replied:
In this driver, the dai[] is statically allocated already, that said, let me add the platform parameter and .id to make it explicit and consistent with other DAI links.

yongzhi1 replied:
Updated https://github.com/thesofproject/linux/commit/0bccc6157ecf26648a5d639c83b1a3575ef0777d

ver 5: add platform to SND_SOC_DAILINK_REG(), also add .id=0 to make it explicit set.

yongzhi1 commented:
Sorry @plbossart , final update https://github.com/thesofproject/linux/commit/268857c917d2ee70fe9c340bc3828a0edb70f0e2

Google caught a build error for "make ARCH=arm allmodconfig":
sound/soc/intel/common/soc-acpi-intel-jsl-match.c:44:16: error: expected declaration specifiers or ‘...’ before string constant
 MODULE_LICENSE("GPL v2");
                ^
So I added #include <linux/module.h> in soc-acpi-intel-jsl-match.c to resolve it.

plbossart commented on sound/soc/intel/common/soc-acpi-intel-jsl-match.c:
```diff
@@ -6,16 +6,33 @@
 
+#include <linux/module.h>
 #include <sound/soc-acpi.h>
-#include <sound/soc-acpi-intel-match.h>
```
this doesn't look right. We have the same set of headers included in all soc-acpi....-match files.

So either it's not needed, or it's needed in all cases.

plbossart replied:
I just tested and I see no problems with make ARCH=arm allmodconfig?

yongzhi1 replied:
make ARCH=arm allmodconfig; then you can see in the .config CONFIG_SND_SOC_ACPI_INTEL_MATCH=m, now make a build, it takes quite long on my build machine to finish "make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf -j4".

With x86 build, it does not need #include <sound/soc-acpi-intel-match.h> so I removed it, this header file further includes the module.h, so it does not fail for arm build for all other files under the same dir except this patch.

plbossart replied:
like I said, things work fine for me with ARCH=arm on topic/sof-dev with the upstream code. You must be using an older environment?

yongzhi1 replied:
Sorry for the confusion, the current code on sof-dev is OK, there is no problem.
This patch removed include <sound/soc-acpi-intel-match.h> since v1 which caused the build error for ARM config, so I added the module.h which is the only necessary include.

FYI, https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/1995884/

yongzhi1 replied:
Hi, @plbossart , morning, in rethinking the #include <sound/soc-acpi-intel-match.h>, I feel it's a correct practice to include this header in it's definition file, so added it back in b84cc98.

kv2019i commented:
@plbossart @ranj063 Testing merging with this PR.
